worker-core/src/main/java/com/hpe/caf/worker/core/WorkerApplication.java

```diff
@@ -108,13 +109,37 @@ public void run(final WorkerConfiguration workerConfiguration, final Environment
         ManagedWorkerQueue workerQueue = queueProvider.getWorkerQueue(config, nThreads);
         TransientHealthCheck transientHealthCheck = new TransientHealthCheck();
         WorkerCore core = new WorkerCore(codec, wtp, workerQueue, workerFactory, path, environment.healthChecks(), transientHealthCheck);
-        Runtime.getRuntime().addShutdownHook(new Thread()
-        {
+
+        environment.lifecycle().manage(new Managed() {
+            @Override
+            public void start() {
+                LOG.debug("Worker starting up.");
+
+                initCoreMetrics(environment.metrics(), core);
+                initComponentMetrics(environment.metrics(), config, store, core);
+
+                final GatedHealthProvider gatedHealthProvider = new GatedHealthProvider(workerQueue, core);
+                environment.healthChecks().register("queue", gatedHealthProvider.new GatedHealthCheck("queue", new WorkerHealthCheck(core.getWorkerQueue())));
+                environment.healthChecks().register("configuration", gatedHealthProvider.new GatedHealthCheck("configuration", new WorkerHealthCheck(config)));
+                environment.healthChecks().register("store", gatedHealthProvider.new GatedHealthCheck("store", new WorkerHealthCheck(store)));
+                environment.healthChecks().register("worker", gatedHealthProvider.new GatedHealthCheck("worker", new WorkerHealthCheck(workerFactory)));
+                environment.healthChecks().register("transient", gatedHealthProvider.new GatedHealthCheck("transient", new WorkerHealthCheck(transientHealthCheck)));
+            }
             @Override
-            public void run()
-            {
-                LOG.debug("Shutting down");
+            public void stop() {
+                LOG.debug("Worker stop requested, allowing in-progress tasks to complete.");
                 workerQueue.shutdownIncoming();
+                while (!wtp.isIdle()) {
+                    try {
+                        //The grace period will expire and the process killed so no need for time limit here
+                        LOG.trace("Awaiting the Worker Thread Pool to become idle, {} tasks in the backlog.", 
+                                wtp.getBacklogSize());
+                        Thread.sleep(1000);
+                    } catch (final InterruptedException e) {
+                        throw new RuntimeException(e);
```

Comment:
I know it's not going to make a lot of difference in this case but you should be resetting the interrupt flag before re-throwing.

```suggestion
                        Thread.currentThread().interrupt();
                        throw new RuntimeException(e);
```